Added command line completion for CLI

maniac-en opened the pull request:
I've also added docs/COMPLETIONS.md which explains how to add shell
completion for bash, zsh and fish shell manually!

Note: This commit doesn't affect the working of quick-lint-js.

Fixes issue: #188

maniac-en commented:
I have read and hereby agree to quick-lint-js' Contributor License Agreement (CLA-v1.md).

maniac-en commented:
recheck

strager commented on docs/COMPLETIONS.md:
```diff
@@ -0,0 +1,49 @@
+### Shell completions
```
Nit: I'd move this file into the `completions` directory and call it `README.md`.

strager reviewed (changes requested):
Blocking: At least the zsh completion seems to be broken.

I typed `quick-lint-js ` then pressed tab. I was presented with a menu of *options*, but I expected to see a menu of *files*.

What I see after pressing tab (with my zsh settings):
```
$ quick-lint-js --exit-fail-on
--exit-fail-on           -- Fail with a non-zero exit code if any of these errors are f
--help            -h     -- Print help message
--lsp-server      --lsp  -- Run quick-lint-js in LSP server mode
--output-format          -- Format to print feedback where FORMAT is one of: gnu-like (
--version         -v     -- Print version information
--vim-file-bufnr         -- Select a vim buffer for outputting feedback
```

What I expect after pressing tab:
```
$ quick-lint-js CMakeLists.txt
CMakeLists.txt  build-debug/    copyrite        hello.js        test/
COPYING         build-fuzz/     default.nix     hello.mjs       tools/
Formula/        build-profile/  dist/           make*           vendor/
README.md       cmake/          docs/           plugin/         website/
benchmark/      completions/    fuzz/           po/
build/          configure*      fuzz-corpus/    src/
```

Blocking: `--output-format` doesn't seem to complete in zsh:

1. If I type `quick-lint-js --out` then press tab, zsh fills in `quick-lint-js --output-format `. This is fine.
2. If I then press tab again, zsh rings a bell and shows no completion options. This is wrong. I expect zsh to present me with a menu of choices.

(Also, ideally, after pressing tab in step #1, zsh would fill in `quick-lint-js --output-format=` and present a menu of choices. Maybe this isn't possible or typical, so I'm fine with the user needing to press tab again.)

maniac-en commented on completions/_quick-lint-js:
```diff
@@ -0,0 +1,22 @@
+#compdef _quick-lint-js quick-lint-js
```
This is not a comment, though I don't see the difference in working of script even if we redact `#`.

I've two reasons to use it as it is:
1. [zsh docs](http://zsh.sourceforge.net/Doc/Release/Completion-System.html)
> If the number of completion files changes, compinit will recognise this and produce a new dump file. However, if the name of a function or the arguments in the first line of a #compdef function (as described below) change, it is easiest to delete the dump file by hand so that the next time compinit will re-create it.

2. [zsh-users HowTo](https://github.com/zsh-users/zsh-completions/blob/master/zsh-completions-howto.org#telling-zsh-which-function-to-use-for-completing-a-command)
> The first line of a completion function file can look something like this:
> 
> #compdef foobar
> 
> This tells zsh that the file contains code for completing the foobar command. This is the format that you will use most often for the first line, but you can also use the same file for completing several different functions [..SNIP..]

maniac-en commented on completions/_quick-lint-js:
```diff
@@ -0,0 +1,22 @@
+#compdef _quick-lint-js quick-lint-js
```
It can be named `quick-lint-js.zsh` but it's suggested to use `_quick-lint-js` according to  [zsh-users HowTo](https://github.com/zsh-users/zsh-completions/blob/master/zsh-completions-howto.org#telling-zsh-which-function-to-use-for-completing-a-command). I'm ok with changing it if you are!

> Completion functions for commands are stored in files with names beginning with an underscore _

maniac-en replied:
Filename kept as `_quick-lint-js` and the file header has now following comments:
> This script is used to provide command line completion for
> quick-lint-js when using zsh shell

maniac-en commented on completions/_quick-lint-js:
```diff
@@ -0,0 +1,22 @@
+#compdef _quick-lint-js quick-lint-js
```
I've added the copyright comments in all completion scripts.

Please update `tools/check-file-copyrights`

strager replied:
Unless you want to maintain copyright for your contribution, the copyright comments should say Matthew Glazar, not Shivam Mehta. They should also say 2020 and have two spaces between the year and my name, exactly like other files.

maniac-en replied:
I don't plan to maintain copyright for my contribution, so I'll change it as specified!

maniac-en commented on completions/_quick-lint-js:
```diff
@@ -0,0 +1,22 @@
+typeset -A opt_args
+
+function _quick-lint-js() {
+	integer ret=1
```
`Changes implemented!`

maniac-en commented on completions/quick-lint-js.bash:
```diff
@@ -0,0 +1,19 @@
+} &&
+	complete -F _quick-lint-js quick-lint-js
+
+# vim: ft=bash:
```
No it's not necessary for this file, the file extension is clear enough already. I added it earlier as it was necessary in other two files, but I've changed it now.

`Changes implemented!`

maniac-en commented on docs/COMPLETIONS.md:
```diff
@@ -0,0 +1,49 @@
+### Shell completions
+
+To get automatic completions for quick-lint-js's options you can install the provided shell completions.
```
`TODO`

maniac-en replied:
`CMakeLists` implemented!

maniac-en commented:
Fixed
> I typed quick-lint-js then pressed tab. I was presented with a menu of options, but I expected to see a menu of files.

Also implemented
> (Also, ideally, after pressing tab in step `#1`, zsh would fill in quick-lint-js --output-format= and present a menu of choices. Maybe this isn't possible or typical, so I'm fine with the user needing to press tab again.)

strager reviewed (commented):
> Also implemented
> 
> > (Also, ideally, after pressing tab in step `#1`, zsh would fill in quick-lint-js --output-format= and present a menu of choices. Maybe this isn't possible or typical, so I'm fine with the user needing to press tab again.)

This doesn't seem to be implemented, at least on zsh version 5.8. Tab completes `--output-format=` with no menu of choices. Pressing tab again has no effect (i.e. no menu).

maniac-en commented:
> This doesn't seem to be implemented, at least on zsh version 5.8. Tab completes `--output-format=` with no menu of choices. Pressing tab again has no effect (i.e. no menu).

Latest commit should fix this.

Along with that I've also added functionality to ignore incompatible parameters:
>Example:
>	`--lsp or --lsp-server` won't be suggested if `--exit-fail-on`
>	or `--output-format` is present

strager commented on completions/.gitignore:
```diff
@@ -0,0 +1,5 @@
+CMakeCache.txt
```
Delete this .gitignore file. (See my blocking comment in completions/README.md.)

maniac-en replied:
`File deleted`

strager commented on completions/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+# See end of file for extended copyright information.
+
+cmake_minimum_required(VERSION 3.10)
+project(quick-lint-js-completions)
```
Why make a `project`? Let's inherit the `project(quick-lint-js)` from `/CMakeLists.txt`.

strager commented on completions/README.md:
````diff
@@ -1,12 +1,18 @@
 
-## Zsh
+```sh
+cmake .
+make
+sudo make install
````
Blocking: We shouldn't provide instructions on *only* building and installing completion files. We should install completion files as part of the CLI build. This is how the manual (`quick-lint-js.1`) is installed, for example.

(The manual installation instructions below are good to have.)

strager commented on completions/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+
+install(
+  FILES _quick-lint-js
+  DESTINATION "/usr/share/zsh/site-functions"
```
Blocking: We should use `${DATAROOTDIR}` from `GNUInstallDirs` instead of hard-coding `/usr/share`. https://cmake.org/cmake/help/v3.10/module/GNUInstallDirs.html

See docs/CMakeLists.txt for a similar example.

maniac-en replied:
`Implemented`

strager commented on completions/_quick-lint-js:
```diff
@@ -0,0 +1,58 @@
+        "($I1 $I3)--output-format=[Format to print feedback where FORMAT is one of\: gnu-like (default if omitted), vim-qflist-json]:format:->format"
+        "($I1)--vim-file-bufnr=[Select a vim buffer for outputting feedback]")
+
+        _arguments -C -S -s "${args[@]}" '*:file:_files' && ret=0
```
quick-lint-js' command line option parser does not support option stacking. Therefore, `-s` shouldn't be given to `_arguments`.

strager commented on completions/_quick-lint-js:
```diff
@@ -0,0 +1,58 @@
+        args+=(
+        "(- *)"{-h,--help}"[Print help message]"
+        "(- *)"{-v,--version}"[Print version information]"
+        "($I1 $I2 --lsp --lsp-server)"{--lsp,--lsp-server}"[Run quick-lint-js in LSP server mode]"
+        "($I1 $I3)--exit-fail-on=[Fail with a non-zero exit code if any of these errors are found (default\: \"all\")]"
+        "($I1 $I3)--output-format=[Format to print feedback where FORMAT is one of\: gnu-like (default if omitted), vim-qflist-json]:format:->format"
+        "($I1)--vim-file-bufnr=[Select a vim buffer for outputting feedback]")
```
The outcome of the exclusions seems a bit weird:

* `quick-lint-js --lsp --<TAB>` will suggest `--exit-fail-on` and `--vim-file-bufnr`. Neither of these options work with `--lsp`.
* `quick-lint-js --output-format=gnu-like --<TAB>` will suggest  `--exit-fail-on` and `--vim-file-bufnr`. `--help` and `--version` also make sense to suggest (in my opinion).

I think the following changes should make it complete how I expect:
```suggestion
        "(--exit-fail-on --lsp --lsp-server --vim-file-bufnr)"{--lsp,--lsp-server}"[Run quick-lint-js in LSP server mode]"
        "(--lsp --lsp-server)--exit-fail-on=[Fail with a non-zero exit code if any of these errors are found (default\: \"all\")]"
        "(--lsp --lsp-server --output-format)--output-format=[Format to print feedback where FORMAT is one of\: gnu-like (default if omitted), vim-qflist-json]:format:->format"
        "(--lsp --lsp-server)--vim-file-bufnr=[Select a vim buffer for outputting feedback]")
```

maniac-en replied:
`Implemented`

strager commented on completions/_quick-lint-js:
```diff
@@ -0,0 +1,53 @@
+        "(--lsp --lsp-server)--exit-fail-on=[Fail with a non-zero exit code if any of these errors are found (default\: \"all\")]"
+        "(--lsp --lsp-server --output-format)--output-format=[Format to print feedback where FORMAT is one of\: gnu-like (default if omitted), vim-qflist-json]:format:->format"
+        "(--lsp --lsp-server)--vim-file-bufnr=[Select a vim buffer for outputting feedback]")
+
```
If `--lsp-server` is on the command line, files shouldn't be completed.

Similarly, if a file is on the command line, `--lsp-server` shouldn't be completed.

strager commented on completions/_quick-lint-js:
```diff
@@ -0,0 +1,53 @@
+        args+=(
+        "(- *)"{-h,--help}"[Print help message]"
+        "(- *)"{-v,--version}"[Print version information]"
+	"(--exit-fail-on --lsp --lsp-server --vim-file-bufnr)"{--lsp,--lsp-server}"[Run quick-lint-js in LSP server mode]"
```
If `--lsp-server` is on the command line, `--output-format` should not be completed.

maniac-en replied:
implemented in next commit!

strager commented on completions/quick-lint-js.bash:
```diff
@@ -0,0 +1,39 @@
+        COMPREPLY=()
+        cur="${COMP_WORDS[COMP_CWORD]}"
+        prev="${COMP_WORDS[COMP_CWORD-1]}"
+        opts="-h --help -v --version --lsp --lsp-server --exit-fail-on --output-format --vim-file-bufnr"
```
Instead of suggesting both `-h` and `--help`, let's suggest only `--help`. I don't want to clutter the completion menu.

Similarly with `-v` and `--version`, and `--lsp` and `--lsp-server`.

strager replied:
Is it possible to include a `=` after `--exit-fail-on` and `--vim-file-bufnr` and `--output-format`?

strager commented on completions/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+
+install(
+  FILES quick-lint-js.fish
+  DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/fish/completions"
```
This looks like the wrong installation location for fish completions, according to my reading of this document:

> The completions shipped with fish, usually installed in /usr/share/fish/completions
> https://fishshell.com/docs/current/completions.html

I think fish/vendor_completions.d/ is a better installation location:

> A directory for third-party software vendors to ship their own completions for their software, usually /usr/share/fish/vendor_completions.d

strager replied:
fish completions were autogenerated from the `man` page on my machine. This is because my fish installation didn't look in /usr/share/fish/completions; it instead looked in /usr/share/fish/vendor_completions.d.

maniac-en replied:
Changed in next commit

strager reviewed (changes requested):
Blocking: CMakeLists.txt needs `add_subdirectory(completions)`. Otherwise, completions/CMakeLists.txt is unused.

maniac-en commented on completions/quick-lint-js.fish:
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env sh
```
I agree, shebang's not needed! Next commit removes shebangs in all completions!

maniac-en commented on completions/quick-lint-js.fish:
```diff
@@ -0,0 +1,31 @@
+# You should have received a copy of the GNU General Public License
+# along with quick-lint-js.  If not, see <https://www.gnu.org/licenses/>.
+
+# vim: ft=sh:
```
Keeping the modeline, shebang removed!

maniac-en commented on completions/_quick-lint-js:
```diff
@@ -0,0 +1,53 @@
+# You should have received a copy of the GNU General Public License
+# along with quick-lint-js.  If not, see <https://www.gnu.org/licenses/>.
+
+# vim: ft=zsh:
```
Removed

maniac-en commented:
As per my findings from fish docs and other fish completions, it behaves differently when it comes to appending `=` to an option unlike bash and zsh. Adding a submenu for `--output-format` in case of fish gives a completion like this (notice the space after `=`)

![13-05-2021_03:17](https://user-images.githubusercontent.com/45049766/118048347-ca47bf80-b399-11eb-908e-6cb36326b7c3.png)

It looks ugly in my opinion, and I'm not sure even functional/valid. On the other hand, it works properly for bash and zsh.
It's your call now if we should keep it or not. :)

strager commented:
> As per my findings from fish docs and other fish completions, it behaves differently when it comes to appending `=` to an option unlike bash and zsh. Adding a submenu for `--output-format` in case of fish gives a completion like this (notice the space after `=`)
> 
> ![13-05-2021_03:17](https://user-images.githubusercontent.com/45049766/118048347-ca47bf80-b399-11eb-908e-6cb36326b7c3.png)
> 
> It looks ugly in my opinion, and I'm not sure even functional/valid. On the other hand, it works properly for bash and zsh.
> It's your call now ~if we should keep it or not.~ about where to proceed with fish completion regarding this concern. :)

Inserting the `=` is incorrect in this case. Remove it.

strager commented on completions/quick-lint-js.fish:
```diff
@@ -1,13 +1,11 @@
+complete -c quick-lint-js -l version -d 'Print version information' -f
+complete -c quick-lint-js -l lsp-server= -d 'Run quick-lint-js in LSP server mode' -r
+complete -c quick-lint-js -l exit-fail-on= -d 'Fail with a non-zero exit code if any of these errors are found (default: "all")' -r
+complete -c quick-lint-js -l output-format= -d 'Format to print feedback where FORMAT is one of: gnu-like (default if omitted), vim-qflist-json' -xa 'gnu-like vim-qflist-json'
```
Blocking: Please fix completion by removing the `=`s.

strager reviewed (changes requested):
Aside from the `Blocking` comment, this looks good to me!
